src/eve/utils.py

```diff
@@ -39,38 +44,135 @@
 from boltons.typeutils import classproperty  # noqa: F401
 
-from .type_definitions import DELETE, NOTHING
-from .typingx import Any, Callable, Iterable, Iterator, List, Optional, Type, Union
+from .type_definitions import NOTHING
+from .typingx import (
+    Any,
+    AnyCallable,
+    Callable,
+    Dict,
+    Iterable,
+    Iterator,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+)
+
 
+try:
+    import cytoolz as toolz
+except ModuleNotFoundError:
+    import toolz
 
-def filter_map(
-    func: Callable[..., Any], iterable: Iterable[Any], *, delete_sentinel: Any = DELETE
-) -> Iterator[Any]:
-    """Mapping function supporting elimination of items.
 
-    Args:
-        iterable: iterable object to be processed.
-        delete_sentinel: sentinel object which marks items to be deleted from the
-            the output (note that comparison is made by identity NOT by value).
+def isinstancechecker(type_info: Union[Type, Iterable[Type]]) -> Callable[[Any], bool]:
+    """Return a callable object that checks if operand is an instance of `type_info`.
 
-    Notes:
-        The default `delete_sentinel` value (:data:`eve.DELETE`) can also be
-        accessed as :data:`filter_map.DELETE`.
+    Examples:
+        >>> checker = isinstancechecker((int, str))
+        >>> checker(3)
+        True
+        >>> checker('3')
+        True
+        >>> checker(3.3)
+        False
 
     """
-    for item in iterable:
-        result = func(item)
-        if result is not delete_sentinel:
-            yield result
 
+    types: Tuple[Type, ...] = tuple()
+    if isinstance(type_info, type):
+        types = (type_info,)
+    elif not isinstance(type_info, tuple) and is_collection(type_info):
+        types = tuple(type_info)
+    else:
+        types = type_info  # type:ignore  # it is checked at run-time
+
+    if not isinstance(types, tuple) or not all(isinstance(t, type) for t in types):
+        raise ValueError(f"Invalid type(s) definition: '{types}'.")
 
-#: Shortcut to the global DELETE sentinel value
-filter_map.DELETE = DELETE  # type: ignore
+    return lambda obj: isinstance(obj, types)
+
+
+def attrchecker(name: str) -> Callable[[Any], bool]:
+    """Return a callable object that checks if operand has `name` attribute.
+
+    Examples:
+        >>> from collections import namedtuple
+        >>> Point = namedtuple('Point', ['x', 'y'])
+        >>> point = Point(1.0, 2.0)
+        >>> checker = attrchecker('x')
+        >>> checker(point)
+        True
+
+        >>> checker = attrchecker('z')
+        >>> checker(point)
+        False
+
+    """
+    if not isinstance(name, str):
+        raise ValueError(f"Invalid attribute name: '{name}'.")
+    return lambda obj: hasattr(obj, name)
 
 
-def get_item(obj: Any, key: Any, default: Any = NOTHING) -> Any:
-    """Similar to :func:`operator.getitem()` accepting a default value."""
+def sattrgetter(*names: str, default: Any = NOTHING) -> Callable[[Any], Any]:
```

Comment:
What's the `s`?

Reply:
The `s` was to follow the `shash` naming to not overlap with the builtin function, but it does not really make sense in this case. Renamed and documented.